### PR TITLE
[Scheduler][Stability] Adding random jitter to address duplicate task ids occurrence

### PR DESCRIPTION
**Pull request by gradientsky**
*Description of changes:*
Added random jitter delay for task id generation - this naturally makes simultaneous state updates/requests less likely.

*Tasks start logs*

Task with id 4 was scheduled multiple times. The problem relatively easy to reproduce without this change on ubuntu, 8-core machine.
``` plain
Scheduling Task (task_id: 0,
	fn: <function lgb_trial at 0x7fa83ace9b00>,
	args: {args: {'util_args': {'dataset_train_filename': 'dataset_train.bin', 'dataset_val_filename': 'dataset_val.b.., config: {'feature_fraction': 1.0, 'learning_rate': 0.1, 'min_data_in_leaf': 20, 'num_leaves': 31}, reporter: DistStatusReporter, },
	resource: DistributedResource(
	Node = Remote REMOTE_ID: 0,
	<Remote: 'inproc://<ip_here>/18386/1' processes=1 threads=8, memory=33.19 GB>
	nCPUs = 1, CPU_IDs = {[0]}))


Scheduling Task (task_id: 1,
	fn: <function lgb_trial at 0x7fa83ace9b00>,
	args: {args: {'util_args': {'dataset_train_filename': 'dataset_train.bin', 'dataset_val_filename': 'dataset_val.b.., config: {'feature_fraction': 0.9247339772148495, 'learning_rate': 0.06592805898944631, 'min_data_in_leaf': 2.., reporter: DistStatusReporter, },
	resource: DistributedResource(
	Node = Remote REMOTE_ID: 0,
	<Remote: 'inproc://<ip_here>/18386/1' processes=1 threads=8, memory=33.19 GB>
	nCPUs = 1, CPU_IDs = {[1]}))


Scheduling Task (task_id: 2,
	fn: <function lgb_trial at 0x7fa83ace9b00>,
	args: {args: {'util_args': {'dataset_train_filename': 'dataset_train.bin', 'dataset_val_filename': 'dataset_val.b.., config: {'feature_fraction': 0.8695569099758053, 'learning_rate': 0.08051659770615258, 'min_data_in_leaf': 2.., reporter: DistStatusReporter, },
	resource: DistributedResource(
	Node = Remote REMOTE_ID: 0,
	<Remote: 'inproc://<ip_here>/18386/1' processes=1 threads=8, memory=33.19 GB>
	nCPUs = 1, CPU_IDs = {[2]}))


Scheduling Task (task_id: 3,
	fn: <function lgb_trial at 0x7fa83ace9b00>,
	args: {args: {'util_args': {'dataset_train_filename': 'dataset_train.bin', 'dataset_val_filename': 'dataset_val.b.., config: {'feature_fraction': 0.8507991265255086, 'learning_rate': 0.07345453930070389, 'min_data_in_leaf': 2.., reporter: DistStatusReporter, },
	resource: DistributedResource(
	Node = Remote REMOTE_ID: 0,
	<Remote: 'inproc://<ip_here>/18386/1' processes=1 threads=8, memory=33.19 GB>
	nCPUs = 1, CPU_IDs = {[3]}))


Scheduling Task (task_id: 4,
	fn: <function lgb_trial at 0x7fa83ace9b00>,
	args: {args: {'util_args': {'dataset_train_filename': 'dataset_train.bin', 'dataset_val_filename': 'dataset_val.b.., config: {'feature_fraction': 0.9516841572012867, 'learning_rate': 0.0699298010777774, 'min_data_in_leaf': 10.., reporter: DistStatusReporter, },
	resource: DistributedResource(
	Node = Remote REMOTE_ID: 0,
	<Remote: 'inproc://<ip_here>/18386/1' processes=1 threads=8, memory=33.19 GB>
	nCPUs = 1, CPU_IDs = {[4]}))


Scheduling Task (task_id: 5,
	fn: <function lgb_trial at 0x7fa83ace9b00>,
	args: {args: {'util_args': {'dataset_train_filename': 'dataset_train.bin', 'dataset_val_filename': 'dataset_val.b.., config: {'feature_fraction': 0.8320992088927449, 'learning_rate': 0.009188210797555645, 'min_data_in_leaf': .., reporter: DistStatusReporter, },
	resource: DistributedResource(
	Node = Remote REMOTE_ID: 0,
	<Remote: 'inproc://<ip_here>/18386/1' processes=1 threads=8, memory=33.19 GB>
	nCPUs = 1, CPU_IDs = {[5]}))


Scheduling Task (task_id: 6,
	fn: <function lgb_trial at 0x7fa83ace9b00>,
	args: {args: {'util_args': {'dataset_train_filename': 'dataset_train.bin', 'dataset_val_filename': 'dataset_val.b.., config: {'feature_fraction': 0.9774783624448347, 'learning_rate': 0.1255406602615921, 'min_data_in_leaf': 14.., reporter: DistStatusReporter, },
	resource: DistributedResource(
	Node = Remote REMOTE_ID: 0,
	<Remote: 'inproc://<ip_here>/18386/1' processes=1 threads=8, memory=33.19 GB>
	nCPUs = 1, CPU_IDs = {[6]}))


Scheduling Task (task_id: 7,
	fn: <function lgb_trial at 0x7fa83ace9b00>,
	args: {args: {'util_args': {'dataset_train_filename': 'dataset_train.bin', 'dataset_val_filename': 'dataset_val.b.., config: {'feature_fraction': 0.9323156247870876, 'learning_rate': 0.05933944114295586, 'min_data_in_leaf': 1.., reporter: DistStatusReporter, },
	resource: DistributedResource(
	Node = Remote REMOTE_ID: 0,
	<Remote: 'inproc://<ip_here>/18386/1' processes=1 threads=8, memory=33.19 GB>
	nCPUs = 1, CPU_IDs = {[7]}))


Scheduling Task (task_id: 4,
	fn: <function lgb_trial at 0x7fa83ace9b00>,
	args: {args: {'util_args': {'dataset_train_filename': 'dataset_train.bin', 'dataset_val_filename': 'dataset_val.b.., config: {'feature_fraction': 0.9516841572012867, 'learning_rate': 0.0699298010777774, 'min_data_in_leaf': 10.., reporter: DistStatusReporter, },
	resource: DistributedResource(
	Node = Remote REMOTE_ID: 0,
	<Remote: 'inproc://<ip_here>/18386/1' processes=1 threads=8, memory=33.19 GB>
	nCPUs = 1, CPU_IDs = {[4]}))


```

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by szha** [comment]
Job PR-753-1 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-753/1/index.html

**Comment by mseeger** [comment]
Can you explain why this works? I'd expect task_id to be created by an entity that exists only once.

**Review by mseeger — APPROVED** [review]

**Comment by gradientsky** [comment]
> Can you explain why this works? I'd expect task_id to be created by an entity that exists only once.

When scheduler starts, it immediately allocates multiple sub-processes to schedule task (8 cores == 8 processes). The way how those processes exchange the information is through `mp.Value` in a shared memory. A Value increment for the counter is not an atomic operation and apparently python state synchronization between threads is not very good at isolating things within an `mp.Lock` boundaries or some CPU caching coming into play. Short story short: it fails at replicating the value between calls so the ids get not unique. 

One of the common ways to combat this kind of racing condition is to add some random start jitter so the processes start slightly delay - this way the lock is not as congested at the startup and likelihood of race condition drops significantly.
